### PR TITLE
#415 Move ScopeDepth from ParseState to ParseGraph.

### PR DESCRIPTION
**Pull request by mvanaken**
See also issue #415.

When de scopeDepth is stored within the ParseState, then calling `withOrder` with a different parseGraph, should actually reset the scopeDepth. Since it is dependendant on the parseGraph it makes more sense to store it there.

### CODE REVIEW

File: core/src/test/java/io/parsingdata/metal/AutoEqualityTest.java [hunk 1]
```diff
@@ -437,4 +436,11 @@ public void basicNoHashCollisions(final Object object, final Object same, final
         }
     }
 
+    private static void assertEquals(final Object o, final Object object) {
```
Comment (mvanaken) [inline core/src/test/java/io/parsingdata/metal/AutoEqualityTest.java:439]: Decided to override the assertEquals methods, so we get more meaningful error messages when a test fails. You will now see which objects causes a test to fail, which previously could only be discovered by debugging the test, which takes much more time.

File: core/src/main/java/io/parsingdata/metal/data/ParseGraph.java [hunk 2]
```diff
@@ -47,52 +48,61 @@ public class ParseGraph extends ImmutableObject implements ParseItem {
...
+            return new ParseGraph(head.asGraph().closeBranch(token), tail, definition, true, newScopeDepth);
+        }
+        if (!head.getDefinition().equals(token)) {
+            throw new IllegalStateException("Cannot close branch with token that does not match its head token.");
+        }
+        if (newScopeDepth != 0) {
+            throw new IllegalStateException("Cannot close parse graph that has a non zero scopeDepth.");
         }
-        return new ParseGraph(head, tail, definition, false);
+        return new ParseGraph(head, tail, definition);
```
Comment (jvdb) [inline core/src/main/java/io/parsingdata/metal/data/ParseGraph.java:105]: Nice change, to call the other constructor here (to make the `false` implicit).

File: core/src/main/java/io/parsingdata/metal/data/ParseGraph.java [hunk 3]
```diff
@@ -140,6 +150,7 @@ public boolean equals(final Object obj) {
             && Objects.equals(head, ((ParseGraph)obj).head)
             && Objects.equals(tail, ((ParseGraph)obj).tail)
             && Objects.equals(branched, ((ParseGraph)obj).branched)
+            && Objects.equals(scopeDepth, ((ParseGraph)obj).scopeDepth)
```
Comment (jvdb) [inline core/src/main/java/io/parsingdata/metal/data/ParseGraph.java:153]: The `scopeDepth` is essentially _derivable_ right? Maybe it should be considered cached then as well just like `size`?

File: core/src/main/java/io/parsingdata/metal/data/ParseState.java [hunk 4]
```diff
@@ -42,63 +42,61 @@ public class ParseState extends ImmutableObject {
     public final Source source;
     public final ImmutableList<ImmutablePair<Token, BigInteger>> iterations;
     public final ImmutableList<ParseReference> references;
-    public final int scopeDepth;
 
-    public ParseState(final ParseGraph order, final ParseValueCache cache, final Source source, final BigInteger offset, final ImmutableList<ImmutablePair<Token, BigInteger>> iterations, final ImmutableList<ParseReference> references, final int scopeDepth) {
```
Comment (jvdb) [inline core/src/main/java/io/parsingdata/metal/data/ParseState.java:47]: Cleans up `ParseState` a lot to move `scopeDepth` to `ParseGraph`!

File: core/src/test/java/io/parsingdata/metal/data/ParseGraphTest.java [hunk 5]
```diff
@@ -236,4 +246,130 @@ public void testCurrent() {
         assertFalse(EMPTY.addBranch(NONE).current().isPresent());
     }
 
+    public static Stream<Arguments> scopeDepthTest() {
```
Comment (jvdb) [inline core/src/test/java/io/parsingdata/metal/data/ParseGraphTest.java:249]: Wow, amazing work on these extensive tests!

**Review by jvdb — APPROVED** [review]

File: core/src/main/java/io/parsingdata/metal/data/ParseGraph.java (hunk 3 above)
Comment (mvanaken) [reply core/src/main/java/io/parsingdata/metal/data/ParseGraph.java:153]: With normal usage, it should be derivable. However, it is possible to create a token that alternates its scope delimiter. This makes it possible to have two identitcal ParseGraphs with different scope depths. This is only theoretical and I do not know why you would create a token like that, but for robustness it is better to keep this here. Other minor reason to keep it here is so we can test equality for the ParseGraph with the AutoEqualityTest.

**Pull request merged**